### PR TITLE
LibC+Ports: Add glib version 2.69.0

### PR DESCRIPTION
**Pull request by kennethmyhra**
This PR adds a port of glib. It feels a bit rough since we are missing a lot of functionality it requires. 

IPV6 support has been disabled.

Since I could not figure out how to use environment variables in the meson cross compilation files there are two of them, one for the i686 and one for the x86_64 target.

There is a README for the patches in the `patches/` directory with brief explanations.

The following methods has been stubbed out in LibC:
- `fnmatch.h`:
  - `fnmatch`
- `resolv.h`:
  - `res_query`
- `xattr.h`:
  - `getxattr`
  - `lgetxattr`
  - `fgetxattr`
  - `setxattr`
  - `lsetxattr`
  - `fsetxattr`
  - `listxattr`
  - `llistxattr`
  - `flistxattr`

image

### CODE REVIEW

File: Ports/glib/package.sh [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+depends=libiconv
+useconfigure=true
+configopts="--cross-file ../cross_file.txt"
+files="https://gitlab.gnome.org/GNOME/glib/-/archive/2.69.0/glib-${version}.tar.gz glib-${version}.tar.gz 35af83aedf34b96f1e99ed9c995c50960509f504750fb4def2440a9406b00a95"
+auth_type=sha256
+
+configure()
+{
+    run meson _build $configopts
+}
```
Comment (linusg) [inline Ports/glib/package.sh:13]: Please stick to the existing style

```suggestion
configure() {
    run meson _build $configopts
}
```

File: Ports/glib/patches/fix-gio-meson.build.patch [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
++if host_system not in ['windows', 'android', 'serenity']
+   # Don't check for C_IN on Android since it does not define it in public
+   # headers, we define it ourselves wherever necessary
+   if not cc.compiles('''#include <sys/types.h>
+@@ -34,7 +34,7 @@ endif
+ 
+ network_libs = [ ]
+ network_args = [ ]
++if host_system not in ['windows', 'serenity']
+-if host_system != 'windows'
```
Comment (linusg) [inline Ports/glib/patches/fix-gio-meson.build.patch:19]: `+` before `-`? did `git`/`diff` do that?

```suggestion
-if host_system != 'windows'
+if host_system not in ['windows', 'serenity']
```

File: Userland/Libraries/LibC/fnmatch.cpp [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+ */
+
+#include <AK/Format.h>
+#include <fnmatch.h>
+
+int fnmatch(char const*, char const*, int)
+{
+    dbgln("FIXME: Implement fnmatch()");
+    return 0;
+}
```
Comment (linusg) [inline Userland/Libraries/LibC/fnmatch.cpp:14]: Please configure your editor to add a newline at the end of files (CI will fail due to this)

**Review by linusg — CHANGES_REQUESTED** [review]
with so many and, more importantly, temporary patches, please add a readme explaining each, so that reasoning about their eventual removal becomes easier. see python or cmake for inspiration.

**Review by linusg — COMMENTED** [review]
> Could not figure out how to use environment variables in the cross_file.txt so it is currently hard-coded to target build i686

Can we just provide two and conditionally set `configopts` in `configure()` based on `SERENITY_ARCH`?

File: Ports/glib/patches/fix-gio-meson.build.patch (hunk 2 above)
Comment (alimpfard) [reply Ports/glib/patches/fix-gio-meson.build.patch:19]: A suggestion for making these patches easier:
- make a separate clean copy of the thing you're trying to port into dir A
- initialise a git repo in A and just `git add -A`
- change stuff in A, commit into there, `git format-patch <hash of the first commmit>` and move all the generated patches to your `Ports/port/patches` dir

with this flow, you can avoid messing with patches and just deal with your own changes in the nicer UI that git has :)

Comment (kennethmyhra) [reply Ports/glib/patches/fix-gio-meson.build.patch:19]: That's a great suggestion, thanks!

**Comment by kennethmyhra** [comment]
Should be good for another review now I think

**Comment by kennethmyhra** [comment]
Setting it to draft for now, I'll fix up the rest of the clang format issues and make sure there are no other issues tomorrow

File: Ports/glib/package.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env -S bash ../.port_include.sh
+port=glib
+version=2.69.0
+depends="libiconv libffi zlib gettext"
+useconfigure=true
+configopts="--cross-file ../cross_file-$SERENITY_ARCH.txt"
```
Comment (linusg) [inline Ports/glib/package.sh:6]: yep that's perfect :+1:

**Review by linusg — APPROVED** [review]

**Comment by kennethmyhra** [comment]
Clang format issues fixed. Also tested with a simple hello world in SerenityOS, see screenshot.

**Pull request merged**